### PR TITLE
Fixed compilation with msvc

### PR DESCRIPTION
**Pull request by arves100**
Under Microsoft Visual C++ compiler, the arguments that starts with - does not exists. (They are exclusive to GCC)
There was also a mistake in include_directories for GLEW.

-O3 is not necessary (counterpart is /Ox) since CMake can Debug/Release builds can be switched inside Vissual Studio
-Wno-errors is disabled by default (/WX-)
narrowing flags are enabled with /W4, default one in CMake is /W3
funroll loops should be enabled when you do a Release build (with optimization)

EDIT: The second commit fixes the compilation issue I've run into visual studio.
- Wrapped "int main" into SDL_main (used by library SDL2_Main) because MSVC may get confused without having WinMain, SDL2 allows to keep compatibility with WinMain (windows specific entry point) and other application by using SDL_main. (I remember that could also be fixed by specifing /SUBSYSTEM:CONSOLE in the linker properties, but it seems like it's not working right)
- Removed Variable Array Length, they are not supported in Visual C++.

### CODE REVIEW

File: src/dmg/apu.cpp [hunk 1]
```diff
@@ -584,10 +584,10 @@ void dmg_audio_callback(void* _apu, u8 *_stream, int _length)
...
 	if(config::use_stereo) { length /= 2; }
 
-	s16 channel_1_stream[length];
-	s16 channel_2_stream[length];
-	s16 channel_3_stream[length];
-	s16 channel_4_stream[length];
+	s16* channel_1_stream = new s16[length];
+	s16* channel_2_stream = new s16[length];
+	s16* channel_3_stream = new s16[length];
+	s16* channel_4_stream = new s16[length];
```
Comment (shonumi) [inline src/dmg/apu.cpp]: This part is kinda ugly. I guess MSVC doesn't like switching to C99... Anyway, something like this is preferred:

    std::vector<s16> channel_1_stream(length);
    std::vector<s16> channel_2_stream(length);
    etc...

This avoids manual memory management and gets rid of having to call **delete**

File: src/dmg/apu.cpp [hunk 2]
```diff
@@ -643,4 +643,9 @@ void dmg_audio_callback(void* _apu, u8 *_stream, int _length)
 			stream[index + 1] = out_sample;
 		}
 	} 
+
+	delete channel_4_stream;
+	delete channel_3_stream;
+	delete channel_2_stream;
+	delete channel_1_stream;
```
Comment (shonumi) [inline src/dmg/apu.cpp]: With std::vector, these aren't necessary. See above comment.

File: src/gba/apu.cpp [hunk 3]
```diff
@@ -365,12 +365,13 @@ void agb_audio_callback(void* _apu, u8 *_stream, int _length)
 	s16* stream = (s16*) _stream;
 	int length = _length/2;
 
-	s16 channel_1_stream[length];
-	s16 channel_2_stream[length];
-	s16 channel_3_stream[length];
-	s16 channel_4_stream[length];
-	s16 dma_a_stream[length];
-	s16 dma_b_stream[length];
```
Comment (shonumi) [inline src/gba/apu.cpp]: Same as above comments. If MSVC doesn't support Variable Length Arrays, C++ vectors are preferred to using **new** and **delete**.

File: src/sgb/core.cpp [hunk 4]
```diff
@@ -935,13 +935,13 @@ void SGB_core::dump_obj(int obj_index) { }
 void SGB_core::dump_bg(int bg_index) { }
 
 /****** Grabs the OBJ palette ******/
-u32* SGB_core::get_obj_palette(int pal_index) { }
+u32* SGB_core::get_obj_palette(int pal_index) { return nullptr; }
```
Comment (shonumi) [inline src/sgb/core.cpp]: Just NULL for now, until GBE+ moves to C++11 or newer. I don't know when/if that'll ever happen.

File: src/sgb/core.cpp [hunk 5]
```diff
@@ -935,13 +935,13 @@ void SGB_core::dump_obj(int obj_index) { }
...
 
 /****** Grabs the BG palette ******/
-u32* SGB_core::get_bg_palette(int pal_index) { }
+u32* SGB_core::get_bg_palette(int pal_index) { return nullptr; }
```
Comment (shonumi) [inline src/sgb/core.cpp]: Same as previous comment.

**Comment by shonumi** [comment]
Thanks for the contribution!

I just have a few things to point out:

1) I'd rather not use **new** and **delete** for arrays. I try to avoid manually managing memory. C++ vectors are preferred. Variable Length Arrays (VLAs) are supported by gcc/g++, but not MSVC apparently. C++ vectors works best for this situation.

2) Using NULL is preferable for now instead of nullptr. If and when GBE+ changes over to C++11 or newer, I'll change them all at once. Might seem kinda silly to others, but I had a specific plan to make C++11 changes, and I'd like to make sure I don't mix and match NULL and nullptr until that time.

3) When you make the above changes, be sure to [squash your multiple commits into a single commit](https://coderwall.com/p/5_cjhw/squash-your-commits-with-git-rebase)

File: src/main.cpp [hunk 6]
```diff
@@ -14,7 +14,9 @@
 #include "nds/core.h"
 #include "common/config.h"
 
-int main(int argc, char* args[]) 
+#include <SDL2/SDL_main.h>
+
+int main(int argc, char** args)
```
Comment (shonumi) [inline src/main.cpp]: Try to keep this as:
    int main(int argc, char *argv[])

**Comment by arves100** [comment]
- Changed VLA to std::vectors
- Fixed linking issues with gbe_plus_qt
- Forced /SUBSYSTEM:CONSOLE when generating exe
- Changed nullptr to NULL.

Note: I didn't use std::vector.data() function since you clarified that gbe+ doesn't use C++11.
Second note: I've not added any resource file, Windows uses the default icon for GBE+ (when you're trying to open the application, they have the default icon), but this does not affect the task bar icon (since it's setted up by SDL2 or Qt)

**Comment by shonumi** [comment]
Looks good! Let me check some things when I get back to my main computer in a few hours, then I'll merge it. Good work!

**Pull request merged**